### PR TITLE
C++: Overridable isSource on DefaultTaintTracking

### PR DESCRIPTION
**Pull request by MathiasVP**
When using `DefaultTaintTracking` we create a configuration that extends `TaintTrackingConfiguration` and define appropriate sinks. However, there is no way to restrict the sources to be anything other than anything that is considered user input.

This seems to be the source of the slowdown in https://github.com/github/codeql/pull/4784: when running `cpp/cgi-xss` on
`openjdk` (which has the largest performance slowdown of about 16%), there are many sources that are user inputs and flow to a sink, but the source isn’t restricted by anything other than “it’s user input” until the where clause of the query, at which point we filter them all away.

This PR adds support for an overridable `isSource` predicate when extending `TaintTrackingConfiguration`. To avoid breaking the API, the default implementation of `isSource` is consistent with what we have on `main` currently.

CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1629/

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll [hunk 1]
```diff
@@ -46,21 +46,23 @@ predicate predictableOnlyFlow(string name) {
 
 private DataFlow::Node getNodeForSource(Expr source) {
   isUserInput(source, _) and
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:48]: This predicate is still called from `taintedWithoutGlobals`. Should it be?

**Review by jbj — COMMENTED** [review]
LGTM except one question.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:48]: Good point. It probably shouldn't.

**Comment by MathiasVP** [comment]
> CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1629/

Performance looks good. The regression in Wireshark doesn't show up in https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1630/ (which also includes this PR) so that's just noise.

**Review by jbj — APPROVED** [review]

**Pull request merged**